ls1intum/Artemis
# [IGNORE] Code-Editor/Resolve CheckoutConflicts in client

### Checklist

### Motivation and Context
If the user of a programming-exercise is making commits simultaneously in the code-editor and locally from git, it can happen under specific circumstances that the the repository enters a CheckoutConflict state.
In some instances, after a reload of the code-editor that causes another pull of the repo, the conflict resolves itself, however in other instances the repo can enter a state that is not resolvable from the code-editor nor the local git repo, but only by logging onto the server and manually resolving the conflicts in the checked out repo.

It would be great, if the student could reset the checkout conflict from within the ArTEMiS client. 

See this stackoverflow post for more info on the checkout conflict: https://stackoverflow.com/questions/18050220/eclipse-egit-checkout-conflict-with-files-egit-doesnt-want-to-continue

### Description


We no longer just log the CheckoutConflict exception but let it bubble up from the pull action to the respective endpoints.
The endpoints then set a 409 (Conflict) to the client.

The implementation in the client is tricky, as all repository endpoints are concerned, but we don't want to manually catch the error in every endpoint and let the conflict state be transferred to the code-editor container.
Because of this, the repository singleton services need to carry an error state for the given participation that the code-editor can use.

This implementation is blocked by https://github.com/ls1intum/ArTEMiS/pull/381.

### Steps for Testing


1. Pull git repo locally
2. Change line in code, stash, commit
3. Change same line in artemis editor, save
4. Push local commit
5. Reload Artemis Editor
=> CheckoutConflictException, Code-Editor is not usable anymore
6. The reset repository button should appear, the save and submit buttons should disappear
7. Open the reset modal by clicking on the reset button
8. Reset the repo by confirming
=> Code-Editor should work again as expected with the state of the last commit


### Screenshots



![CheckoutConflict_1](https://user-images.githubusercontent.com/28230611/58746483-eed15800-845e-11e9-925c-c2af185bc3dd.png)

![CheckoutConflict_2](https://user-images.githubusercontent.com/28230611/58746484-f1cc4880-845e-11e9-9deb-e4dd4260e265.png)
